src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java

```diff
@@ -1337,25 +1340,76 @@ public Type constructorType() {
+                csym.appendUniqueTypeAttributes(constr.getRawTypeAttributes().stream()
+                        .filter(anno -> anno.position.type == TargetType.METHOD_RETURN)
+                        .collect(List.collector()));
+
+                ListBuffer<VarSymbol> params = new ListBuffer<>();
+                List<Type> argtypes = constructorType().getParameterTypes();
+                if (!enclosingType().hasTag(NONE)) {
                     argtypes = argtypes.tail;
+                    params = params.prepend(new VarSymbol(PARAMETER, make.paramName(0), enclosingType(), csym));
+                }
+                if (constr.params != null) {
+                    for (VarSymbol p : constr.params) {
+                        VarSymbol param = new VarSymbol(PARAMETER | p.flags(), p.name, argtypes.head, csym);
+                        param.appendAttributes(p.getRawAttributes());
+                        param.appendUniqueTypeAttributes(p.getRawTypeAttributes());
+                        params.add(param);
+                        argtypes = argtypes.tail;
+                    }
                 }
+                csym.params = params.toList();
+
+                // now lets create the tree to make sure that it is created according to our needs
+                methodDecl = make.MethodDef(constrParams(csym),
+                        typeParams(csym.type.getTypeArguments()), csym, csym.type, null);
+                return csym;
             }
-            csym.params = params.toList();
-            return csym;
+            return constructorSymbol;
         }
 
+        // helper methods
+            private List<JCVariableDecl> constrParams(MethodSymbol csym) {
+                ListBuffer<JCVariableDecl> params = new ListBuffer<>();
+                if (csym.params != null) {
+                    for (VarSymbol param : csym.params) {
+                        /* the compiler will issue an error if a type annotation is applied to, for example, `java.lang.String`
+                         * but it won't if the type annotation is applied to `String`, at this point all types will be represented
+                         * in its fully unfolded form so we need to go back to the short form if type annotations are present for a
+                         * given parameter
+                         */
+                        if (!param.getRawTypeAttributes().isEmpty()) {
+                            params.append(make.VarDef(param, make.Ident(param.type.tsym), null));
+                        } else {
+                            params.append(make.VarDef(param,null));
+                        }
+                    }
+                }
+                return params.toList();
+            }
+
+            private List<JCTypeParameter> typeParams(List<Type> typarams) {
+                ListBuffer<JCTypeParameter> tparams = new ListBuffer<>();
+                final AtomicInteger paramIndex = new AtomicInteger(0);
```

Comment:
The AtomicInteger is a big weird (it is the first time it is used in javac,  believe). I wonder if using some alternative (like an array) would not fit better.